File: src/api/discussion-recipients.ts
import { RequestFN } from "~/core/request-function";
import { decodeDiscussionRecipient } from "~/decoders/discussion-recipient";
import { Discussion, DiscussionRecipient, SessionHandle, TabLocation } from "~/models";

/**
 * Fetches the recipients of the discussion.
 *
 * A recipient is someone who is part of the discussion.
 * They don't have to send a message to be considered as a recipient.
 */
export const discussionRecipients = async (session: SessionHandle, discussion: Discussion): Promise<Array<DiscussionRecipient>> => {
  const request = new RequestFN(session, "SaisiePublicMessage", {
    _Signature_: { onglet: TabLocation.Discussions },

    donnees: {
      message: { N: discussion.participantsMessageID },
      estPublicParticipant: true,
      estDestinatairesReponse: false
    }
  });

  const response = await request.send();
  return response.data.donnees.listeDest.V.map(decodeDiscussionRecipient);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/discussion-recipients.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeDiscussionRecipient
import ink.literate.pawnote.models.Discussion
import ink.literate.pawnote.models.DiscussionRecipient
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun discussionRecipients(
    session: SessionHandle,
    discussion: Discussion
): List<DiscussionRecipient> {
  val request =
      RequestFN(
          session.information,
          "SaisiePublicMessage",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Discussions.code) }

                putJsonObject("donnees") {
                  putJsonObject("message") { put("N", discussion.participantsMessageID) }
                  put("estPublicParticipant", true)
                  put("estDestinataireReponse", false)
                }
              }))

  val response = request.send()
  return Json.parseToJsonElement(response.data)
      .jsonObject["donnees"]!!
      .jsonObject["listeDest"]!!
      .jsonObject["V"]!!
      .jsonArray
      .map { decodeDiscussionRecipient(it.jsonObject) }
}
